runtime/ops/otel.rs

```diff
@@ -0,0 +1,582 @@
+  type Sender<T: Debug + Send> = BatchMessageChannelSender<T>;
+
+  fn batch_message_channel<T: Debug + Send>(
+    &self,
+    capacity: usize,
+  ) -> (Self::Sender<T>, Self::Receiver<T>) {
+    let (batch_tx, batch_rx) = tokio::sync::mpsc::channel::<T>(capacity);
+    (batch_tx.into(), batch_rx.into())
+  }
+}
+
+#[derive(Debug)]
+pub struct BatchMessageChannelSender<T: Send> {
+  sender: tokio::sync::mpsc::Sender<T>,
+}
+
+impl<T: Send> From<tokio::sync::mpsc::Sender<T>>
+  for BatchMessageChannelSender<T>
+{
+  fn from(sender: tokio::sync::mpsc::Sender<T>) -> Self {
+    Self { sender }
+  }
+}
+
+impl<T: Send> opentelemetry_sdk::runtime::TrySend
+  for BatchMessageChannelSender<T>
+{
+  type Message = T;
+
+  fn try_send(
+    &self,
+    item: Self::Message,
+  ) -> Result<(), opentelemetry_sdk::runtime::TrySendError> {
+    self.sender.try_send(item).map_err(|err| match err {
+      tokio::sync::mpsc::error::TrySendError::Full(_) => {
+        opentelemetry_sdk::runtime::TrySendError::ChannelFull
+      }
+      tokio::sync::mpsc::error::TrySendError::Closed(_) => {
+        opentelemetry_sdk::runtime::TrySendError::ChannelClosed
+      }
+    })
+  }
+}
+
+pub struct BatchMessageChannelReceiver<T> {
+  receiver: tokio::sync::mpsc::Receiver<T>,
+}
+
+impl<T> From<tokio::sync::mpsc::Receiver<T>>
+  for BatchMessageChannelReceiver<T>
+{
+  fn from(receiver: tokio::sync::mpsc::Receiver<T>) -> Self {
+    Self { receiver }
+  }
+}
+
+impl<T> Stream for BatchMessageChannelReceiver<T> {
+  type Item = T;
+
+  fn poll_next(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+  ) -> Poll<Option<Self::Item>> {
+    self.receiver.poll_recv(cx)
+  }
+}
+
+fn otel_create_globals(
+  config: OtelConfig,
+  op_state: &mut OpState,
+) -> anyhow::Result<()> {
+  // Parse the `OTEL_EXPORTER_OTLP_PROTOCOL` variable. The opentelemetry_*
+  // crates don't do this automatically.
+  // TODO(piscisaureus): enable GRPC support.
+  let protocol = match env::var("OTEL_EXPORTER_OTLP_PROTOCOL").as_deref() {
+    Ok("http/protobuf") => Protocol::HttpBinary,
+    Ok("http/json") => Protocol::HttpJson,
+    Ok("") | Err(env::VarError::NotPresent) => {
+      return Ok(());
+    }
+    Ok(protocol) => {
+      return Err(anyhow!(
+        "Env var OTEL_EXPORTER_OTLP_PROTOCOL specifies an unsupported protocol: {}",
+        protocol
+      ));
+    }
+    Err(err) => {
+      return Err(anyhow!(
+        "Failed to read env var OTEL_EXPORTER_OTLP_PROTOCOL: {}",
+        err
+      ))
+    }
+  };
+
+  // Define the resource attributes that will be attached to all log records.
+  // These attributes are sourced as follows (in order of precedence):
+  //   * The `service.name` attribute from the `OTEL_SERVICE_NAME` env var.
+  //   * Additional attributes from the `OTEL_RESOURCE_ATTRIBUTES` env var.
+  //   * Default attribute values defined here.
+  // TODO(piscisaureus): add more default attributes (e.g. script path).
+  let mut resource = Resource::default();
```

Comment:
Can we use `opentelemetry_sdk::resource::SdkProvidedResourceDetector` to detect things from env vars?

Reply:
`Resource::default` will automatically pick up on `OTEL_SERVICE_NAME` and `OTEL_RESOURCE_ATTRIBUTES`.